src/shims/native_lib/mod.rs

```diff
@@ -276,6 +278,147 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
+            ty::Uint(UintTy::U64) => CPrimitive::UInt64(scalar(v)?.to_u64()?).into(),
+            ty::Uint(UintTy::Usize) =>
+                CPrimitive::USize(scalar(v)?.to_target_usize(this)?.try_into().unwrap()).into(),
+            ty::RawPtr(..) => {
+                let ptr = scalar(v)?.to_pointer(this)?;
+                // Pointer without provenance may not access any memory anyway, skip.
+                if let Some(prov) = ptr.provenance {
+                    // The first time this happens, print a warning.
+                    if !this.machine.native_call_mem_warned.replace(true) {
+                        // Newly set, so first time we get here.
+                        this.emit_diagnostic(NonHaltingDiagnostic::NativeCallSharedMem { tracing });
+                    }
+
+                    this.expose_provenance(prov)?;
+                };
+
+                // This relies on the `expose_provenance` in the `visit_reachable_allocs` callback
+                // below to expose the actual interpreter-level allocation.
+                CPrimitive::RawPtr(std::ptr::with_exposed_provenance_mut(ptr.addr().bytes_usize()))
+                    .into()
+            }
+            // For ADTs, create an FfiType from their fields.
+            ty::Adt(adt_def, args) => {
+                let strukt = this.adt_to_ffitype(v.layout.ty, *adt_def, args)?;
+
+                // Copy the raw bytes backing this arg.
+                let bytes = match v.as_mplace_or_imm() {
+                    either::Either::Left(mplace) => {
+                        // We do all of this to grab the bytes without actually
+                        // stripping provenance from them, since it'll later be
+                        // exposed recursively.
+                        let ptr = mplace.ptr();
+                        // Make sure the provenance of this allocation is exposed;
+                        // there must be one for this mplace to be valid at all.
+                        // The interpreter-level allocation itself is exposed in
+                        // visit_reachable_allocs.
+                        this.expose_provenance(ptr.provenance.unwrap())?;
+                        // Then get the actual bytes.
+                        let id = this
+                            .alloc_id_from_addr(
+                                ptr.addr().bytes(),
+                                mplace.layout.size.bytes_usize().try_into().unwrap(),
+                                /* only_exposed_allocations */ true,
+                            )
+                            .unwrap();
+                        let ptr_raw = this.get_alloc_bytes_unchecked_raw(id)?;
+                        // SAFETY: We know for sure that at ptr_raw the next layout.size bytes
+                        // are part of this allocation and initialised. They might be marked as
+                        // uninit in Miri, but all bytes returned by the isolated allocator are
```

Comment:
FWIW before https://github.com/rust-lang/miri/pull/4475 this argument was wrong on macOS where we didn't use the isolated allocator.

And in general there's no need to be so specific.
```suggestion
                        // uninit in Miri, but all bytes returned by `MiriAllocBytes` are
```